src/middleware.ts

```diff
@@ -259,43 +296,60 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
-
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+  // The state that was loaded from the storage and is set via set() within this
+  // middleware would later be overridden with the initial state by create(). To
+  // avoid this we merge the state from localStorage into the initial state. This
+  // way the state gets overridden with the state loaded from storage.
+  // TODO: find a better solution for this
```

Comment:
Hmm... this is only for sync. Not typed. It seems really TODO.

Reply:
True, the type is missing. I guess it would be
```js
  let stateFromStorage: S | void
```